### PR TITLE
MEPTS-1075 | TB1 - COMPLETION CASCADE

### PR DESCRIPTION
**Pull request by helderjosue**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java [hunk 1]
```diff
@@ -711,44 +771,53 @@ public CohortDefinition getINHStartA3(int encounterType, int profilaxiaIsoniazid
...
+            + " encounter e ON p.patient_id = e.patient_id"
+            + " INNER JOIN"
+            + " obs o ON e.encounter_id = o.encounter_id"
+            + " INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id    "
+            + " WHERE"
+            + " p.voided = 0 AND e.voided = 0 AND o.voided = 0"
+            + "    AND e.encounter_type IN (${encounterTypes})"
+            + " AND (o.concept_id = ${23985} AND o.value_coded = ${656})   "
+            + " AND (o2.concept_id = ${165308} AND o2.value_coded = ${1256})   "
+            + "    AND o2.obs_datetime < :endDate "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]:
```suggestion
            + " AND (o2.concept_id = ${165308} AND  o2.value_coded = ${1256}   "
            + "    AND o2.obs_datetime < :endDate) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java [hunk 2]
```diff
@@ -625,42 +669,50 @@ public CohortDefinition getINHStartA1(
...
-            + "    AND e.encounter_type = ${6}"
-            + "    AND o.concept_id = ${6122}"
-            + "    AND o.value_coded = ${1256}"
-            + "    AND e.encounter_datetime < :endDate"
-            + "    AND e.location_id = :location";
+            + " AND e.encounter_type = ${encounterTypes}"
+            + " AND (o.concept_id = ${23985}"
+            + " AND o.value_coded = ${23954})"
+            + " AND (o2.concept_id = ${165308} AND o2.value_coded IN (${states})) "
+            + " AND o2.obs_datetime < :endDate"
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]:
```suggestion
            + " AND (o2.concept_id = ${165308} AND  o2.value_coded IN (${states}) "
            + " AND o2.obs_datetime < :endDate)"
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java [hunk 3]
```diff
@@ -960,8 +1029,8 @@ public CohortDefinition getINHStartA6(
             + "  AND e.encounter_type = ${60} "
             + "  AND (o.concept_id = ${23985} AND o.value_coded IN (${656} , ${23982})) "
             + "  AND (o2.concept_id = ${23987} AND o2.value_coded IN (${1256} , ${1705})) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]:
```suggestion
            + "  AND (o2.concept_id = ${23987} AND  o2.value_coded IN (${1256} , ${1705}) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java [hunk 4]
```diff
@@ -960,8 +1029,8 @@ public CohortDefinition getINHStartA6(
...
-            + "  AND e.encounter_datetime < :endDate  "
-            + "  AND e.location_id = :location";
+            + "  AND o2.obs_datetime < :endDate  "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]:
```suggestion
            + "  AND o2.obs_datetime < :endDate)  "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java [hunk 5]
```diff
@@ -1217,56 +1286,781 @@ public CohortDefinition get3HPStartC3(
...
+            + "AND o.concept_id = ${23985} AND o.value_coded IN(${23954}, ${23984})   "
+            + "AND e.location_id = :location       "
+            + "AND e.encounter_datetime < :endDate "
+            + "AND p.voided = 0 AND e.voided = 0 AND o.voided = 0   "
+            + ") AS regimeTPT    "
+            + "ON regimeTPT.patient_id = p.patient_id   "
+            + "WHERE e.encounter_type = ${6}   "
+            + "AND (o.concept_id = ${23985} AND o.value_coded = ${23954}) "
+            + "AND (o2.concept_id = ${165308} AND o2.value_coded = ${1256})   "
+            + "OR (o.concept_id = ${1719} AND o.value_coded IN (${23954},${165307})) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]:
```suggestion
            + "AND ((o2.concept_id = ${165308} AND  o2.value_coded = ${1256})   "
            + "OR (o.concept_id = ${1719} AND  o.value_coded IN (${23954},${165307}))) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java [hunk 6]
```diff
@@ -1217,56 +1286,781 @@ public CohortDefinition get3HPStartC3(
...
+            + "   AND o.concept_id = ${23985} AND o.value_coded IN(${23954},${23984})   "
+            + "   AND e.location_id = :location   "
+            + "   AND e.encounter_datetime BETWEEN DATE_SUB(regimeTPT.encounter_datetime, INTERVAL 4 MONTH) AND regimeTPT.encounter_datetime   "
+            + "   AND p.voided = 0 AND e.voided = 0 AND o.voided = 0   "
+            + ") ) AS regimeTPT    "
+            + "ON regimeTPT.patient_id = p.patient_id   "
+            + "WHERE e.encounter_type = ${53}   "
+            + " AND o.concept_id = ${23985} AND o.value_coded = ${23954} "
+            + " AND e.location_id = :location  AND o2.concept_id = ${165308} AND o2.value_coded = ${1267} "
+            + " AND DATEDIFF(o2.obs_datetime, regimeTPT.encounter_datetime) >= 86 AND DATEDIFF(o2.obs_datetime, regimeTPT.encounter_datetime) <= 365 AND o2.obs_datetime <= :endDate"
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]:
```suggestion
            + " AND e.location_id = :location  AND (o2.concept_id = ${165308} AND  o2.value_coded = ${1267} "
            + " AND DATEDIFF(o2.obs_datetime, regimeTPT.encounter_datetime) >= 86 AND DATEDIFF(o2.obs_datetime, regimeTPT.encounter_datetime) <= 365 AND o2.obs_datetime <= :endDate)"
```

**Review by psmatsinhe — CHANGES_REQUESTED** [review]
@helderjosue Please have a look

**Review by psmatsinhe — CHANGES_REQUESTED** [review]
@helderjosue Please have a look at this comments

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java (hunk 6 above)
Comment (helderjosue) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTCompletionCohortQueries.java]: I pushed a commit with all suggestions applied.

**Review by psmatsinhe — APPROVED** [review]
Looks good to me

**Pull request merged**